Fix: bug when requesting Jetpack sites after plugin installation

adalpari opened the pull request:
## Description
This PR fixes a bug happening after Jetpack plugin installation. The bug was caused by not taking into account the current Application Password credential inside the `FETCH_SITES` action. We were preventing this on the Application Password setup by using a different call, but it turned out that it is better to take it into account here as well.
On the other hand, I've modified the authentication call to not mess with the new changes.

## Testing instructions

1. Log into a non-Jetpack self-hosted site
2. Enable Application Password and Jetpack Rest connection features.
3. Set up Application Password
4. Install Jetpack
- [x] Verify the Application Password credentials are still there by checking the DB or checking that the AP card is not popping up.
5. Revoke the credentials in your wp-admin panel (Users -> [USER] -> Application Password)
6. Try to open the Media screen
- [x] Notice you are asked to re-authenticate
7. Re-authenticate
- [x] Verify you can open Media now.

https://github.com/user-attachments/assets/62ca8514-58b6-46b2-962b-997ec65e66bc


https://github.com/user-attachments/assets/ede7ef96-be5f-4b5c-9024-f85ea3262c10


<!--

Consider testing the following *yourself* before requesting a review:

- Compatibility with WordPress.com sites and self-hosted Jetpack sites
- Both portrait and landscape orientations
- Light and dark modes
- Dynamic font sizes: larger, smaller and bold text
- High contrast mode
- Screen reader experience (e.g., Talkback)
- Languages with large words or with letters/accents not frequently used in English
- Right-to-left languages layout
- Multi-tasking: split view and slide over

-->

<!--

Use these emoji to improve the code review

❤️ - Praise. Compliments about great solution.
⛏ - Nitpick️. Not high importance. Usually, short one-liners.
💡 - Idea. Alternative solution or improvement.
❓ - Question. Requires some explanation or more context.
⚠️ - Warning. Should be addressed in this PR or in some follow-up PR.
🛑 - Blocker. Must be addressed before PR gets merged.

-->

adalpari committed:
Preserving Application Password credentials if exist

adalpari committed:
Fixing check condition

adalpari committed:
fixing the single site fetch

adalpari committed:
Merge remote-tracking branch 'origin/trunk' into fix/CMM-690-Bug-when-requesting-jetpack-sites

adalpari committed:
detekt

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

adalpari commented on libs/fluxc/src/main/java/org/wordpress/android/fluxc/store/SiteStore.kt:
```diff
@@ -1495,20 +1504,29 @@ open class SiteStore @Inject constructor(
+                        siteModel.apiRestUsernameIV = freshSiteFromDB.apiRestUsernameIV
+                        siteModel.apiRestPasswordIV = freshSiteFromDB.apiRestPasswordIV
+                        siteModel.wpApiRestUrl = freshSiteFromDB.wpApiRestUrl
+                    }
```
I don't think this is adding better readability

adalpari commented on libs/fluxc/src/main/java/org/wordpress/android/fluxc/store/SiteStore.kt:
```diff
@@ -1553,20 +1571,28 @@ open class SiteStore @Inject constructor(
+                        site.apiRestUsernameIV = siteFromDB.apiRestUsernameIV
+                        site.apiRestPasswordIV = siteFromDB.apiRestPasswordIV
+                        site.wpApiRestUrl = siteFromDB.wpApiRestUrl
+                    }
```
See: https://github.com/wordpress-mobile/WordPress-Android/pull/22154#discussion_r2287657427

adalpari committed:
typo

adalpari committed:
Merge branch 'trunk' into fix/CMM-690-Bug-when-requesting-jetpack-sites

nbradbury commented on libs/fluxc/src/main/java/org/wordpress/android/fluxc/store/SiteStore.kt:
```diff
@@ -1495,20 +1504,29 @@ open class SiteStore @Inject constructor(
-    @Suppress("ForbiddenComment", "SwallowedException")
-    private fun updateSite(siteModel: SiteModel): OnSiteChanged {
+    @Suppress("ForbiddenComment", "SwallowedException", "NestedBlockDepth")
+    private fun updateSite(siteModel: SiteModel, forceApplicationPasswordOverride: Boolean = false): OnSiteChanged {
```
**NP:** I'm not sure if `forceApplicationPasswordOverride` is the right name for this parameter because to me it doesn't describe its purpose well. Perhaps something like `copyApplicationPassword` or `retainApplicationPassword` which defaults to `true`?

adalpari replied:
mmmm, I don't think it's retaining the value but overriding it in the DB instead. I mean, it's taking the credentials from the `siteModel` variable and overriding the ones in the DB if set to true.
Maybe I could add some javadoc to the function so it's a bit more clear?

nbradbury replied:
> Maybe I could add some javadoc to the function so it's a bit more clear?

Yep that would work!

nbradbury commented on libs/fluxc/src/main/java/org/wordpress/android/fluxc/action/SiteAction.java:
```diff
@@ -150,6 +150,8 @@ public enum SiteAction implements IAction {
     @Action(payloadType = SiteModel.class)
     UPDATE_SITE,
+    @Action(payloadType = SiteModel.class)
+    UPDATE_SITE_APPLICATION_PASSWORD,
```
I think this is confusing. It's called `UPDATE_SITE_APPLICATION_PASSWORD` as though it will update the application password from the backend, but what it really does is fetch the site and _not_ retain the application password.

If we keep the name the same, I think we'll want to add a comment stating its purpose for clarity.

nbradbury replied:
Hmm, based on your previous comment perhaps I'm misunderstanding how this works?

adalpari replied:
I'm not sure to follow your perspective. The action is the same as `UPDATE_SITE`, and actually not fetching the site but updating it with the object `SiteModel` provided. `UPDATE_SITE` is retaining the Application Password from the DB because of the same reason as `mobileEditor` or `webEditor` fields. They are not present othewise. 
`FETCH_SITES` is fetching sites, and sharing the same call to the internal `updateSite` function. So, maybe there it is the confusion.

Now, this is why I named it as "override". When override is `false`, we keep the ones from the DB, and when it's `true`, we override the ones in the DB. so, the new action is `UPDATING_SITE` but also `OVERRIDING` the AP.

I see the confusion, though. Let me see if I can simplify the logic, and maybe auto-detect the override cases...

nbradbury replied:
Ah, yes, I was definitely confused about all of that. A comment stating "When override is false, we keep the passwords from the DB, and when it's true, we override the ones in the DB" would help here.

adalpari committed:
Creating independent actions to add and remove Application Password credentials

adalpari committed:
Fixing tests

adalpari committed:
Merge branch 'fix/CMM-690-Bug-when-requesting-jetpack-sites' of https://github.com/wordpress-mobile/WordPress-Android into fix/CMM-690-Bug-when-requesting-jetpack-sites

adalpari commented:
So, from our conversation, I've done the following:

1. Removed `forceApplicationPasswordOverride` for clarity. So, now `updateSite` looks the same as previously but with the AP fixes
2. Removed `UPDATE_SITE_APPLICATION_PASSWORD` for clarity as well and created `UPDATE_APPLICATION_PASSWORD` and `REMOVE_APPLICATION_PASSWORD`, which are way more clear. It was non-trivial to include all the cases in the `updateSite`, especially detecting cases when override was necessary. So, now with these two new options, it looks better imo.

Please, test the AP flow to check that I haven't introduced any regression bug.

https://github.com/user-attachments/assets/15aa9815-0b83-43ff-8756-b366f447013a

https://github.com/user-attachments/assets/da7258ea-fdd7-4426-8c80-be3602479c20

nbradbury commented on libs/fluxc/src/main/java/org/wordpress/android/fluxc/store/SiteStore.kt:
```diff
@@ -1517,6 +1533,57 @@ open class SiteStore @Inject constructor(
+            if (siteFromDB == null) {
+                OnSiteChanged(SiteError(SiteErrorType.INVALID_SITE))
+            } else {
+                // Adding the plain credential will trigger an encruption before adding them into the DB
```
This comment is the same as the one on `updateApplicationPassword`. Is that intentional? Also, a **np:** typo in "encruption."

adalpari replied:
I've removed them. They are not useful anymore. Thanks!

nbradbury reviewed (approved):
These changes look good! I left a minor note about a comment, but feel free to merge when ready :shipit:

adalpari committed:
Removing useless comment
